File: 1. Fundamentals/eval_rpn.py
class Solution:

    def evalRPN(self, tokens: List[str]) -> int:
        def _helper(acc, curr):
            if (curr not in '/+-*'):
                acc.append(int(curr))
            else:
                last = acc.pop()
                second_to_last = acc.pop()
                if(curr == '/'):
                    acc.append(int(second_to_last / last))
                elif(curr == '+'):
                    acc.append(second_to_last + last)
                elif(curr == '-'):
                    acc.append(second_to_last - last)
                else:
                    acc.append(last * second_to_last)
            return acc

        return reduce(_helper, tokens, [])[0]


# is number?
# def _helper(x, y):
#     print(x, y)
#     return x + y
# print(reduce(_helper, [1, 2, 3, 4, 5]))

File: 1. Fundamentals/EvalRPN.js
/**
 * @param {string[]} tokens
 * @return {number}
 * https://leetcode.com/problems/evaluate-reverse-polish-notation/
 */
var evalRPN = function(tokens) {
  return tokens.reduce((acc, curr) => {
    if (isNaN(curr)) {
      let last = Number(acc.pop());
      let secondToLast = Number(acc.pop());
      switch (curr) {
        case '*':
          acc.push(last * secondToLast);
          break;
        case '+':
          acc.push(last + secondToLast);
          break;
        case '/':
          acc.push(Math.trunc(secondToLast / last));
          break;
        case '-':
          acc.push(secondToLast - last);
          break;
      }
    } else acc.push(curr);
    return acc;
  }, [])[0];
};

// console.log(evalRPN(['2', '1', '+', '3', '*']));
// console.log(evalRPN(['4', '13', '5', '/', '+']));
// console.log(
//   evalRPN(['10', '6', '9', '3', '+', '-11', '*', '/', '*', '17', '+', '5', '+'])
// );

// console.log(isNaN('*'));
// console.log('4' * '6');
// console.log(Math.trunc(6 / -132));
